include/seqan3/search/dream_index/interleaved_bloom_filter.hpp

```diff
@@ -714,6 +733,101 @@ class counting_vector : public std::vector<value_t>
 
 };
 
+/*!\brief Manages counting ranges of values for the seqan3::interleaved_bloom_filter.
+ * \attention Calling seqan3::interleaved_bloom_filter::increase_bin_number_to invalidates the counting_agent_type.
+ *
+ * \details
+ *
+ * ### Example
+ *
+ * \include test/snippet/search/dream_index/counting_agent.cpp
+ */
+template <data_layout data_layout_mode>
+template <std::integral value_t>
+class interleaved_bloom_filter<data_layout_mode>::counting_agent_type
+{
+private:
+    //!\brief The type of the augmented seqan3::interleaved_bloom_filter.
+    using ibf_t = interleaved_bloom_filter<data_layout_mode>;
+
+    //!\brief A pointer to the augmented seqan3::interleaved_bloom_filter.
+    ibf_t const * ibf_ptr;
+
+    //!\brief Store a seqan3::interleaved_bloom_filter::membership_agent to call `bulk_contains`.
+    typename ibf_t::membership_agent membership_agent;
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    counting_agent_type() = default; //!< Defaulted.
+    counting_agent_type(counting_agent_type const &) = default; //!< Defaulted.
+    counting_agent_type & operator=(counting_agent_type const &) = default; //!< Defaulted.
+    counting_agent_type(counting_agent_type &&) = default; //!< Defaulted.
+    counting_agent_type & operator=(counting_agent_type &&) = default; //!< Defaulted.
+    ~counting_agent_type() = default; //!< Defaulted.
+
+    /*!\brief Construct a counting_agent_type for an existing seqan3::interleaved_bloom_filter.
+     * \private
+     * \param ibf The seqan3::interleaved_bloom_filter.
+     */
+    counting_agent_type(ibf_t const & ibf) : ibf_ptr(std::addressof(ibf)), membership_agent(ibf)
+    {
+        result_buffer.resize(ibf_ptr->bin_count());
+    };
+    //!\}
+
+    //!\brief Stores the result of count_values().
+    counting_vector<value_t> result_buffer;
+
+    /*!\name Counting
+     * \{
+     */
+    /*!\brief Counts the occurrences in each bin for all values in a range.
+     * \tparam value_range_t The type of the range of values. Must model std::ranges::input_range. The reference type
+     *                       must model std::unsigned_integral.
+     * \param[in] values The range of values to process.
+     *
+     * \attention The result of this function must always be bound via reference, e.g. `auto &` to prevent copying.
+     * \attention Sequential calls to this function invalidate the previously returned reference.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/dream_index/counting_agent.cpp
+     *
+     * ### Thread safety
+     *
+     * Concurrent invocations of this function are not thread safe, please create a seqan3::counting_agent_type for
+     * each thread.
+     */
+    template <std::ranges::range value_range_t>
+    [[nodiscard]] counting_vector<value_t> const & count_values(value_range_t && values) & noexcept
```

Comment:
just a thing to think about:

Since the `memberhip_agent` has the function `bulk_contains` we might want to consider naming this function
`bulk_count`.